offer-taker/taker-order.md

```diff
@@ -100,7 +100,9 @@ event OrderComplete(
     address indexed inbound_tkn,
     address taker,
     uint takerGot, // net amount of outbound tokens received by taker
-    uint takerGave // total amount of inbound tokens sent by taker
+    uint takerGave, // total amount of inbound tokens sent by taker
+    uint penalty, // the total penalty collected by msg.sender as bounty for failing offers
+    uint feePaid // the fee paid by the taker
```

Comment:
@adhusson Lasse and I are discussing whether "taker" and "msg.sender" are the same here?
It seems a bit weird that these two lines distinguish between the two, but maybe there's a good reason?

Reply:
The comments align with the doc/impl. The first paragraph of https://code.mangrove.exchange/MgvDoc.html#mgvoffertaking.sol-market-order states:
> It returns four uints: the total amount of outbound_tkn received, the total amount of inbound_tkn spent, the penalty received by msg.sender (in wei), and the fee paid by the taker (in wei).
This aligns with the implementation in `payTakerMinusFees` which sends fee to vault and the rest to taker, and `sendPenalty` which sends to `msg.sender`.

The msg.sender != taker is related to https://code.mangrove.exchange/MgvDoc.html#mgvoffertakingwithpermit.sol - but I am not sure _when_ this would actually be used, @adhusson , can you elaborate?

Reply:
Yes it is related to the permit feature. This is supposed to mimic the `approve` and `permit` features of ERC20 tokens whereby someone gets the authorization to move funds on your behalf. It lets you give to someone the authorization to buy on Mangrove on your behalf.

From the doc just below your link (I should add an easy way to link to anywhere):
> The delegate version of `marketOrder` is `marketOrderFor`, which takes a `taker` address as additional argument. Penalties incurred by failed offers will still be sent to `msg.sender`, but exchanged amounts will be transferred from and to the `taker`. If the `msg.sender`’s allowance for the given `outbound_tkn`,`inbound_tkn` and `taker` are strictly less than the total amount eventually spent by `taker`, the call will fail.

Why do this? Because a taker may want a smart contract to buy stuff for them, as part of another workflow (see all uses of `transferFrom`). Of course time will tell if this feature actually gets used -- maybe "make for me" will be more useful than this "take for me" feature. But a basic "make for me" is possible without modifying Mangrove, as implemented by Jean's strats.

Why send the penalty to msg.sender? Because the person initiating the transaction is paying for gas and should be compensated for failed offers, not the person spending the funds. Maybe Mangrove should send the penalty to `tx.origin` but 1) this can be reimplemented by having msg.sender bounce the penalty back up to `tx.origin` if desired, within the tx, and 2) Vitalik said "Do NOT assume that tx.origin will continue to be usable or meaningful."